### DISCUSSION TITLE
Changes to deps.clj to aid embedding

### DISCUSSION THREAD

**Discussion by cursive-ghost** [borkdude/deps.clj#88] [Ideas]
I'm planning to embed deps.clj into Cursive, to replace my use of the CLI scripts and calling into t.d.a directly. I've been investigating this, and there are a couple of changes that would make this easier. I'm happy to provide PRs for these if needed, but I thought I'd check first whether the changes are desirable.

Most of these are simple refactorings to break the functionality up a bit. Currently the `-main` function is monolithic, which makes it difficult to access individual pieces of the functionality.

1. Initially I tried just bundling deps.clj and requiring it like any other namespace. That didn't work in Cursive because of the `io/resource` call for obtaining `DEPS_CLJ_VERSION`. `io/resource` uses the thread context class loader by default, and when requiring the ns I can't control that class loader. This might be worth fixing, although I don't think it will be a problem for me in the final version of my code since I'll be loading deps.clj into a ShimDandy shim.
2. When the script needs to download the Clojure tools, I would like to be able to control how that download happens. I'd like to be able to use IntelliJ's HTTP client API, which will automatically use the user's proxy configuration as well as showing progress bars etc. Currently there's no good way to achieve this. I can think of two options for this:
   - have a dynamic var holding a function similar to `clojure-tools-download-java` which is initially set to null. `clojure-tools-jar-download` could then check if that is bound and invoke it as the first download option if so. 
   - allow the script to stop if the download is required, and output the information about what is needed to download so that I can fetch it and then re-invoke the script. That seems a little messy, though.
3. Related to the above, when embedding it would be great to be able to avoid setting system properties like the proxy ones. It would be good to split `jvm-proxy-settings` into two parts, one returning the proxy details to be passed to `clojure-tools-download-java`, and one which actually sets the properties. That could be called within `clojure-tools-download-direct` so the properties are only set if we know we need them in the current process.
4. For Cursive to do its thing I'll need to read in the `.libs` file since the `-Stree` output doesn't have enough information. However, there's no way to figure out the checksum for a particular configuration at the moment. It would be useful to break up the initial part of `-main` into functions that calculate the various requirements (`cache-dir`, `config-paths` etc) and also calculate the checksum so that I can figure out which file from the `cpcache` corresponds to a particular configuration. 
5. **edit:** Also, there are still some calls to `System/getenv` which haven't been indirected through `*getenv-fn*`.

I don't think any of these changes are difficult or particularly dangerous, they're mostly just splitting things apart a bit more than they are now. If they sound ok I'm happy to provide PRs for them.

**Comment by borkdude** [comment]
@cursive-ide Hi Colin, I'd be happy to incorporate these changes. Perhaps we can go through them one by one: post an issue + corresponding PR.

I've also been thinking of creating a `borkdude.deps.api` namespace which is more suited to be used from Clojure, rather than a main invokation. I've been delaying that by introducing dynamic vars, but I haven't really documented that behavior well enough to call that an API.

**Comment by cursive-ghost** [comment]
Sounds good, I'll get on to that today.

**Comment by cursive-ghost** [comment]
One further question: how would you feel about having another dynamic var for subprocesses (classpath generation, POM creation), similar to `*process-fn*`? Something like:
```clojure
(def ^:private ^:dynamic *subprocess-fn* shell-command)
```

So by default the behaviour would be as it is now, and `shell-command` would be called, but that could be customised if required.

I ask because it seems like the subprocess output redirection is not happening as I expected. If there's an error printed (e.g. [this one](https://github.com/clojure/tools.deps.alpha/blob/5b8909bdb072d66c9f643bae50cd2878a0f694c2/src/main/clojure/clojure/tools/deps/alpha/script/make_classpath2.clj#L231)), then it comes out on stderr of the client process, not `*err*`. This means that I can't easily capture this output to be shown in the IntelliJ UI.

If this sounds like a reasonable change, I'll make an issue and send a PR.

**Reply by borkdude** [reply]
OK, let's consider this after the API PR:

https://github.com/borkdude/deps.clj/pull/97

**Reply by borkdude** [reply]
@cursive-ide Can you explain why `*process-fn*` itself is not sufficient?

**Reply by borkdude** [reply]
I think I get it: `*process-fn*` is only used to start clojure, but not for the intermediate steps (calculating the classpath).
Yes, I'm fine with introducing another one for that.

I renamed `*process-fn*` to `*clojure-process-fn*` in the `api` branch and it only receives a map of `{:cmd [...]}` which are the command line arguments to start the java process which starts clojure.

I'm going to do a similar thing for the intermediate process: `*java-process-fn*` maybe?

**Reply by borkdude** [reply]
I named it `*aux-process-fn*` in the `api` branch.
